### PR TITLE
Upgrade to use Node.js 20 and bump npm dependencies

### LINKED ISSUE #2002: Upgrade to Node 20
## Details
This issue is for upgrading the Node version since 16 (current version) has reached the end of it’s [security support period](https://href.li/?https://endoflife.date/nodejs). Since we rely on WP and WC components we should probably follow along with the Gutenberg issue:

https://github.com/WordPress/gutenberg/issues/48588

Their current plan is to jump to Node 20.

This issue will also be helpful to collect any blockers which might be preventing us from upgrading the Node version.

### Is your feature request related to a problem? 

Node 16 (current version) has reached the end of it’s [security support period]

### Describe the solution you'd like 
Update to Node 20

### Why blocked

With the current dependency tree, there are some packages do not / will not support Node.js v20, for example:

1. `@es-joy/jsdoccomment@0.20.1`, a child dependency of `@wordpress/scripts@24.6.0`, [doesn't support v18](https://github.com/es-joy/jsdoccomment/blob/v0.20.1/package.json#L51).
2. This extension currently uses `@woocommerce/components@10.3.0`, but `@woocommerce/components@12.0.0` (the current L-2 WC) [only supports v16](https://github.com/woocommerce/woocommerce/blob/7.8.2/packages/js/components/package.json#L8).

The first example blocks this extension to use v18 for now, and the second one may force it to downgrade to v16 in the future.

### How to check

Since packages don't always accurately reflect the scope of Node.js support for children dependencies, confirming that these blocks have been removed usually requires actually upgrading the dependencies and testing the installation with Node.js v18 to see if it completes successfully.

Until then, another way to get a quick overview of package's Node.js support is `npm view <packageName> engines`, for example:

```bash
> npm view "@woocommerce/components@>=12" engines
@woocommerce/components@12.0.0 { node: '^16.13.1', pnpm: '^7.13.3' }
@woocommerce/components@12.1.0 { node: '^16.14.1', pnpm: '^8.6.5' }
```

## Acceptance criteria

* [ ]  Node engines set to >= 20
* [ ] .nvmrc set to v20
* [ ] App compiles without errors with Node 20
* [ ] JS tests run correctly without errors with Node 20

### PR DESCRIPTION
**Pull request by eason9487**
### Changes proposed in this Pull Request:

Closes #2002

This PR:
- Bump npm dependencies and upgrade to use Node.js v20.
- Fix the build config of Fast Refresh mode.
- Address Sass deprecation warnings.
   - Ref: https://sass-lang.com/documentation/breaking-changes/mixed-decls/
- Turn off a few eslint rules:
   - `react/react-in-jsx-scope` rule
   - `jsdoc/check-line-alignment` rule
   - `import/no-named-as-default` rule
- Resolve eslint errors: (💡 It would be easier to review the changes by individual commits)
- Automatically fixed errors by running `npm run lint:js -- --fix`: [c41ef64](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/c41ef64f1b2a3e3cef675729185d8797fded520b)
- `@wordpress/i18n-text-domain` rule: [cdd9aab](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/cdd9aabfd374206cea3b525875f4733ccf95c8b7)
- `jsdoc/no-undefined-types` rule: [67ef0e2](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/67ef0e2225dd452423708c6c62a23946d8bb0280)
- `import/no-unresolved` rule: [c4db2a3](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/c4db2a3f4dd55075bc06d70e0b98beb5eec2a1b7)
- `no-redeclare` rule: [6997b6f](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/6997b6fa2730e0be5c507e2f556710c4d3c8c066)
- `jsdoc/require-yields-check` rule: [47d1f9f](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/47d1f9f44fe466171456a131d0ec74290d1c7d6f)
- `@typescript-eslint/no-use-before-define` rule: [9886344](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/9886344a7fea97645b420481271d0e5a8952a128)
- `jsdoc/tag-lines` rule: [fccc0cd](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/fccc0cde1a9e21f1ddd9de9c5805895b7b080eae)

❗ Since the jest testing requires a lot of adjustments, it will be handled by #2544. Two of the eslint rule errors are highly related to jest and are therefore handled by #2544 as well.

### Detailed test instructions:

1. Set Node.js to v20
2. Run `npm install` to see if it can finish without errors
3. Check if the following scripts work well
   - `npm run build`
   - `npm run dev`
   - `npm run start`
   - `npm run start:hot`
   - `npm run lint:css`
   - `npm run lint:js`
      - There are two pre-existing warnings that still can not be fixed after this upgrade
      - Errors from the `testing-library/no-unnecessary-act` and `testing-library/no-unnecessary-act` rules are handled separately in #2544
4. Run `node ./node_modules/.bin/bundlewatch` to check if the check works
5. Do some smoke testing

### Changelog entry

> Dev - Upgrade to use Node.js 20 and bump npm dependencies.

### CODE REVIEW

**Comment by puntope** [comment]
> Address Sass deprecation warnings.

I still see the deprecations in both this PR and in https://github.com/woocommerce/google-listings-and-ads/pull/2544

Screenshot 2024-08-21 at 14 04 13

**Comment by eason9487** [comment]
Hi @puntope

> I still see the deprecations in both this PR and in #2544

The SCSS styles that caused deprecations came from `@wordpress/base-styles` and could not be resolved from this repo.

File: js/src/components/contact-information/contact-information-preview-card.scss [hunk 1]
```diff
@@ -13,9 +13,9 @@
 	}
 
 	&__placeholder {
-		@include placeholder;
-
 		display: inline-block;
 		width: 18em;
+
+		@include placeholder;
```
Comment (puntope) [inline js/src/components/contact-information/contact-information-preview-card.scss:19]: Is this change necessary?  I wonder if placeholder includes  `display` or `width` props. Then it will be overrided

File: js/src/components/free-listings/setup-free-listings/index.js [hunk 2]
```diff
@@ -2,6 +2,7 @@
  * External dependencies
  */
 import { useRef } from '@wordpress/element';
+import { Form } from '@woocommerce/components';
```
Comment (puntope) [inline js/src/components/free-listings/setup-free-listings/index.js:5]: Wondering if it's worth it to import the component just for the docs.

File: package.json [hunk 3]
```diff
@@ -171,7 +170,7 @@
 		}
 	},
 	"engines": {
-		"node": "^16 || ^18",
-		"npm": "^8 || ^9"
```
Comment (puntope) [inline package.json:175]: So I assume dependabot now supports

"node": "^20", and "npm": "^10"

?

File: .eslintrc.js [hunk 4]
```diff
@@ -39,19 +40,78 @@ module.exports = {
...
+		// - `import { userEvent } from '@testing-library/user-event';`
+		//   It works but the official documentation also recommends using the default export
+		'import/no-named-as-default': 'off',
 		'jest/expect-expect': [
 			'warn',
 			{ assertFunctionNames: [ 'expect', 'expect[A-Z]\\w*' ] },
 		],
+		// Turn it off temporarily because it involves a lot of re-alignment. We can revisit it later.
+		'jsdoc/check-line-alignment': 'off',
+		// The JS package `tracking-jsdoc` changes the definition of the `@fires` tag.
```
Comment (puntope) [inline .eslintrc.js]: Can you elaborate what do you mean by this? 

 The JS package `tracking-jsdoc` changes the definition of the `@fires` tag.

**Review by puntope — COMMENTED** [review]
Howdy @eason9487 I tested all the scripts, code and did several smoke testing:

- Onboarding
- Edit campaigns
- Create rules in attribute mapping
- Campaign assets

Everything is working fine without errors so far.

I just left some comments regarding one CSS and just to be sure dependabot is compatible with the new node version.

File: js/src/components/contact-information/contact-information-preview-card.scss (hunk 1 above)
Comment (eason9487) [reply js/src/components/contact-information/contact-information-preview-card.scss:19]: > Is this change necessary?

The change fixes a Sass deprecation warning. The explanation can be found in https://sass-lang.com/documentation/breaking-changes/mixed-decls/

image

> I wonder if placeholder includes `display` or `width` props. Then it will be overrided

They don't have. Please refer to:

https://github.com/woocommerce/google-listings-and-ads/blob/fccc0cde1a9e21f1ddd9de9c5805895b7b080eae/js/src/css/abstracts/_mixins.scss#L9-L21

File: js/src/components/free-listings/setup-free-listings/index.js (hunk 2 above)
Comment (eason9487) [reply js/src/components/free-listings/setup-free-listings/index.js:5]: The main difference is the code editor's reference hints will be better, apart from that there shouldn't be any difference. So it should be worth it.

**Without import**

image

**With import**
image

File: package.json (hunk 3 above)
Comment (eason9487) [reply package.json:175]: Thanks for catching this!

I thought dependabot would use paired node and npm versions. Surprisingly, it's using unpaired Node.js 20 and npm 9. Even the npm version claimed by its error message is incorrect, bringing more confusion.

```
Dependabot uses Node.js v20.17.0 and NPM 10.8.2. Due to the engine-strict setting, the update will not succeed.
```

Fixed in https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/b2a3d130e3a54cdb3fed71832962ca2ea6e10468

File: .eslintrc.js (hunk 4 above)
Comment (eason9487) [reply .eslintrc.js]: Adjusted the code comment in https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/c76514babdc1a4f51eab7b390b75aba61083d586. Let me know if it's still not clear enough.

**Comment by eason9487** [comment]
Hi @puntope, thanks for the review! This PR is ready for a new round of code reviews.

File: .eslintrc.js (hunk 4 above)
Comment (puntope) [reply .eslintrc.js]: I will just write "Listing our event names to avoid eslint false alarms"

But not a blocker tho.

**Review by puntope — APPROVED** [review]
✅  LGTM 

Thanks for the changes and all your work here
